src/test/java/io/aiven/kafka/auth/audit/UserOperationsActivityAuditorTest.java

```diff
@@ -0,0 +1,169 @@
+        assertEquals(1, auditor.auditStorage.size());
+        auditor.dump();
+        assertEquals(0, auditor.auditStorage.size());
+
+        verify(session).principal();
+        verify(session).clientAddress();
+        verify(operation).name();
+        verify(resource).resourceType();
+        verify(resource).name();
+
+    }
+
+    @Test
+    void shouldAggregateOperationsForSameUser() throws Exception {
+
+        final Session anotherSession =
+            new Session(principal, InetAddress.getByName("127.0.0.2"));
+
+        final UserOperationsActivityAuditor auditor = createAuditor();
+
+        auditor.addActivity(session, operation, resource, false);
+        auditor.addActivity(session, operation, resource, true);
+        auditor.addActivity(anotherSession, operation, resource, true);
+        assertEquals(2, auditor.auditStorage.size());
+        assertEquals(
+            2,
+            auditor.auditStorage.get(
+                new Auditor.AuditKey(
+                    session.principal(),
+                    session.clientAddress())
+            ).operations.size()
+        );
+        assertEquals(
+            1,
+            auditor.auditStorage.get(
+                new Auditor.AuditKey(
+                    anotherSession.principal(),
+                    anotherSession.clientAddress())
+            ).operations.size()
+        );
+        auditor.dump();
+        assertEquals(0, auditor.auditStorage.size());
+
+        verify(session, times(3)).principal();
+        verify(session, times(3)).clientAddress();
+        verify(operation, times(3)).name();
+        verify(resource, times(3)).resourceType();
+        verify(resource, times(3)).name();
+
+    }
+
+    @Test
+    public void shouldBuildRightLogMessage() throws Exception {
+        final UserOperationsActivityAuditor auditor = createAuditor();
+
+        final UserActivity userActivity = new UserActivity();
+        userActivity.addOperation(new UserOperation(operation, resource, false));
+
+        String auditMessage =
+            auditor.auditMessage(
+                new Auditor.AuditKey(session.principal(), session.clientAddress()),
+                userActivity
+            );
+
+        assertEquals(
+            String.format(
+                "PRINCIPAL_TYPE:PRINCIPAL_NAME (%s) was active since %s: Deny OP_NAME on RESOURCE_TYPE:RESOURCE_NAME.",
+                InetAddress.getLocalHost(), userActivity.activeSince.format(DateTimeFormatter.ISO_INSTANT)
+            ),
+            auditMessage
+        );
+
+        final Operation anotherOperation = mock(Operation.class);
+        when(anotherOperation.name()).thenReturn("ANOTHER_OP_NAME");
+
+        userActivity.addOperation(new UserOperation(anotherOperation, resource, true));
+        auditMessage = auditor.auditMessage(
+            new Auditor.AuditKey(session.principal(), session.clientAddress()),
+            userActivity
+        );
+        assertEquals(
```

Comment:
This is flaky due to unordered nature of `HashSet`. We should allow both variants or ordering between ops in the message string.